src/Foundation/NSUrlSessionHandler.cs

```diff
@@ -215,13 +215,18 @@ InflightData GetInflightData (NSUrlSessionTask task)
 					if (sessionHandler.inflightRequests.TryGetValue (task, out inflight))
 						return inflight;
 
+				// if we did not manage to get the inflight data, we either got an error or have been canceled, lets cancel the task, that will execute DidCompleteWithError
+				task.Cancel ();
 				return null;
 			}
 
 			public override void DidReceiveResponse (NSUrlSession session, NSUrlSessionDataTask dataTask, NSUrlResponse response, Action<NSUrlSessionResponseDisposition> completionHandler)
 			{
 				var inflight = GetInflightData (dataTask);
 
+				if (inflight == null)
+					return;
+
 				try {
 					// ensure that we did not cancel the request, if we did, do cancel the task
 					if (inflight.CancellationToken.IsCancellationRequested) {
```

Comment:
I agree the `inflight` null check is needed everywhere `GetInflightData` is called. 

I also agree that `GetInflightData` is also the best place to handle cancellations.

My point is that it only deals with one case, i.e. there are two sources of cancellation, the ones we get from the native`NSUrlSession*` itself, and the ones from the managed `DidReceiveData `.

The later is only handled in `DidReceiveResponse`, but if that checks is moved into `GetInflightData` then it would be checked every time to OS call us back, which seems better. E.g. should not `DidReceiveData` also check the `DidReceiveData` token ?